iree/samples/static_library/static_library_demo.c

```diff
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// A example of static library loading in IREE. See the README.md for more info.
+// Note: this demo requires artifacts from iree-translate before it will run.
+
+#include <stdio.h>
+
+#include "iree/hal/local/loaders/static_library_loader.h"
+#include "iree/hal/local/task_device.h"
+#include "iree/modules/hal/hal_module.h"
+#include "iree/runtime/api.h"
+#include "iree/samples/static_library/simple_mul_c.h"
+#include "iree/task/api.h"
+#include "iree/vm/bytecode_module.h"
+
+// Compiled static library module here to avoid IO:
+#include "iree/samples/static_library/simple_mul.h"
+
+// A function to create the HAL device from the different backend targets.
+// The HAL device is returned based on the implementation, and it must be
+// released by the caller.
+iree_status_t create_device_with_static_loader(iree_hal_device_t** device) {
+  // Set paramters for the device created in the next step.
+  iree_hal_task_device_params_t params;
+  iree_hal_task_device_params_initialize(&params);
+
+  // Load the statically embedded library
+  const iree_hal_executable_library_header_t** static_library =
+      simple_mul_dispatch_0_library_query(
+          IREE_HAL_EXECUTABLE_LIBRARY_LATEST_VERSION, /*reserved=*/NULL);
+  const iree_hal_executable_library_header_t** libraries[1] = {static_library};
+
+  iree_hal_executable_loader_t* library_loader = NULL;
+  IREE_RETURN_IF_ERROR(iree_hal_static_library_loader_create(
+      IREE_ARRAYSIZE(libraries), libraries, iree_allocator_system(),
+      &library_loader));
+
+  iree_task_executor_t* executor = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_task_executor_create_from_flags(iree_allocator_system(), &executor));
+
+  // Create the device and release the executor and loader afterwards.
+  IREE_RETURN_IF_ERROR(iree_hal_task_device_create(
+      iree_make_cstring_view("dylib"), &params, executor, 1, &library_loader,
+      iree_allocator_system(), device));
+  iree_task_executor_release(executor);
+  iree_hal_executable_loader_release(library_loader);
+
+  return iree_ok_status();
+}
+
+iree_status_t Run() {
+  // Instance configuration (this should be shared across sessions).
+  iree_runtime_instance_options_t instance_options;
+  iree_runtime_instance_options_initialize(IREE_API_VERSION_LATEST,
+                                           &instance_options);
+  iree_runtime_instance_options_use_all_available_drivers(&instance_options);
+  iree_runtime_instance_t* instance = NULL;
+  IREE_RETURN_IF_ERROR(iree_runtime_instance_create(
+      &instance_options, iree_allocator_system(), &instance));
+
+  // Create dylib device with static loader.
+  iree_hal_device_t* device = NULL;
+  IREE_RETURN_IF_ERROR(create_device_with_static_loader(&device),
+                       "create device");
+  iree_vm_module_t* hal_module = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_hal_module_create(device, iree_allocator_system(), &hal_module));
```

Comment:
This should technically use the style like
https://github.com/google/iree/blob/79cf15440a6be860a8821e5cb520d62a8829a216/iree/samples/variables_and_state/main.c#L131-L134

(notice all the `iree_status_is_ok` checks in that file)

Instead of `IREE_RETURN_IF_ERROR`. Otherwise, this can return without calling the cleanup code below, leaking memory
```c

  // Cleanup call and buffers.
  iree_hal_buffer_unmap_range(&mapped_memory);
  iree_hal_buffer_view_release(ret_buffer_view);
  iree_runtime_call_deinitialize(&call);

  // Cleanup session and instance.
  iree_runtime_session_release(session);
  iree_runtime_instance_release(instance);

  return iree_ok_status();
```

Reply:
I think this would be a lot messier:

If it's possible to proceed through the file with a failure at some unknown point previously, it's no longer clear what was actually allocated and therefor needs to be released, versus just a null pointer. I'd need to also surround every release with a null check. The code is also a lot messier as a result. At the very least, I'd just encapsulate this in a macro as well. 

Given that this is a test/sample, it will stop immediately on failure and print an error, dropping all state in the process.